### DISCUSSION TITLE
Home Assistant - Auto register sensor

### DISCUSSION THREAD

**Discussion by hutten0** [mp-se/gravitymon#128] [General]
First, thanks for a nice firmware! Just love to be able to log more than one endpoint!
It seems to be wrong device classes in the template for Home Assistant - Auto register sensor.
Both gravity and tilt has "temperature". I think they shouldn’t have any at all.
A couple of wishes:
-	In tilt and angle, two more digits please, or at least in one of them.
I’m planning to calibrate by fermenting a sugar wash and take measurements along the way.
-	Add Battery % as in “device_class” “battery”

I never managed to succeed with Home Assistant - Auto register sensor, so I made a script in HA to do the job and it’s persistent to.

Cheers

**Comment by mp-se** [comment]
> First, thanks for a nice firmware! Just love to be able to log more than one endpoint! It seems to be wrong device classes in the template for Home Assistant - Auto register sensor. Both gravity and tilt has "temperature". I think they shouldn’t have any at all. A couple of wishes:
> 
> * In tilt and angle, two more digits please, or at least in one of them.
>   I’m planning to calibrate by fermenting a sugar wash and take measurements along the way.
> * Add Battery % as in “device_class” “battery”
> 
> I never managed to succeed with Home Assistant - Auto register sensor, so I made a script in HA to do the job and it’s persistent to.
> 
> Cheers

Thanks for the feedback. 

For the temp I can agree that one more digit would be good to have. I don't really see the need for having 3 digits on the temp since the max accuracy is 0.05C anyway when using 12 digits resolution.

On the gyro I believe the accuracy is 1% so more than 2 digits would not give you any better accuracy anyway.  

On the device classes for HA I agree that that should be changed. I will add that to the backlog.

The way the auto registration works is that you need to run it once when in configuration mode for the entities to register correctly but they will be lost when you restart home assistant. During fermentation it only sends the data parts to reduce the time needed for writing to mqtt. I could add an option so that you can choose if the auto registration should be done at every data transmit. That would guarantee that the values are persisted every time. Let me know what you think ?

**Reply by hutten0** [reply]
Thanks for the reply.
I didn’t ask for more digits for temp, but yes one more doesn’t hurt :-)
It was tilt and angle I was asking for, at least change one of them to 3-4 digits.
Everything must add up in the formula for calculating the gravity where 4 digits really matters.

_“I could add an option so that you can choose if the auto registration should be done at every data transmit.”_
No, not good. Sometimes my iSpindel 8266 struggles sending the payload…

As I mentioned – I solved the auto register in HA with a script like below:

 ```
gravmon_autodiscovery:
  alias: iSpindel MQTT discovery
  sequence:
    - service: mqtt.publish
      data:
        retain: true
        topic: homeassistant/sensor/iSpindel001/gravity/config
        payload: >-
          {"device": {
          "identifiers": ["d3ce67"],
          "name": "iSpindel001",
          "model": "GravityMon",
          "manufacturer": "Magnus Persson",
          "sw_version": "1.2.1" },
          "unit_of_measurement": "SG",
          "value_template": "{% raw %}{{ value|float }}{% endraw %}",
          "state_topic": "gravmon/iSpindel001/gravity",
          "name": "iSpindel001 Gravity",
          "unique_id": "iSpindel001_grav",
          "icon": "mdi:microscope"
          }
...
```

Most important here is under data: retain: true. This way it will be saved by the broker in my case mosquitto, this works even after restarting HA.

Edit:
Oh, forgot about battery % - any comments about that?

Cheers Patrik

**Reply by mp-se** [reply]
Sure, I see your point. I can add more digits to the angle to give that more accuracy. From what I can read the accuracy is 1% so 3 digits would allow show the accuracy. + more digits for temperature.

I can update the battery variable as well.

**Reply by hutten0** [reply]
Great, thanks!

**Reply by mp-se** [reply]
I looked into the battery part and according to the HA documentation that value should be a percentage of the charge. I would suggest that this is added as a new parameter since the existing voltage is aligned with HA sensor types.

**Reply by hutten0** [reply]
Perfect, thanks!

**Reply by hutten0** [reply]
I was looking in the dev branch, and I think there is a typo in service.rst for gravity. retain: false should be true, right?

**Reply by mp-se** [reply]
Yes, that was a typo

**Reply by mp-se** [reply]
You can try the version in the dev branch if you want. I still have more tests to do before I push it to master branch.
